packages/cli/src/commands/hydrogen/env/list.ts

```diff
@@ -0,0 +1,108 @@
+import {Flags} from '@oclif/core';
+import Command from '@shopify/cli-kit/node/base-command';
+import {renderConfirmationPrompt, renderTable} from '@shopify/cli-kit/node/ui';
+import {outputContent, outputToken} from '@shopify/cli-kit/node/output';
+
+import {linkStorefront} from '../link.js';
+import {adminRequest} from '../../../lib/graphql.js';
+import {commonFlags} from '../../../lib/flags.js';
+import {getHydrogenShop} from '../../../lib/shop.js';
+import {getAdminSession} from '../../../lib/admin-session.js';
+import {
+  ListEnvironmentsQuery,
+  ListEnvironmentsSchema,
+} from '../../../lib/graphql/admin/list-environments.js';
+import {getConfig} from '../../../lib/shopify-config.js';
+import {logMissingStorefront} from '../../../lib/missing-storefront.js';
+import {logMissingLink} from '../../../lib/missing-link.js';
+
+export default class List extends Command {
+  static description = 'List the environments on your Hydrogen storefront.';
+
+  static hidden = true;
+
+  static flags = {
+    path: commonFlags.path,
+    shop: commonFlags.shop,
+  };
+
+  async run(): Promise<void> {
+    const {flags} = await this.parse(List);
+    await listEnvironments(flags);
+  }
+}
+
+interface Flags {
+  path?: string;
+  shop?: string;
+}
+
+export async function listEnvironments({path, shop: flagShop}: Flags) {
+  const shop = await getHydrogenShop({path, shop: flagShop});
+  const adminSession = await getAdminSession(shop);
+  const actualPath = path ?? process.cwd();
+  let configStorefront = (await getConfig(actualPath)).storefront;
+
+  if (!configStorefront?.id) {
+    logMissingLink({adminSession});
+
+    const runLink = await renderConfirmationPrompt({
+      message: outputContent`Run ${outputToken.genericShellCommand(
+        `npx shopify hydrogen link`,
+      )}?`.value,
+    });
+
+    if (!runLink) {
+      return;
+    }
+
+    await linkStorefront({path, shop: flagShop, silent: true});
+  }
+
+  configStorefront = (await getConfig(actualPath)).storefront;
+
+  if (!configStorefront) {
+    return;
+  }
+
+  const result: ListEnvironmentsSchema = await adminRequest(
+    ListEnvironmentsQuery,
+    adminSession,
+    {
+      id: configStorefront.id,
+    },
+  );
+
+  const hydrogenStorefront = result.hydrogenStorefront;
+
+  if (!hydrogenStorefront) {
+    logMissingStorefront({adminSession, storefront: configStorefront});
+    return;
+  }
+
+  const rows = hydrogenStorefront.environments
+    .filter(({type}) => type !== 'PREVIEW')
```

Comment:
Why do we need to filter out preview types? 🤔

Reply:
This is a great question! I'm filtering it because it looks a little odd in the table because it doesn't have a branch or URL:

```
Name        Branch              URL
──────────  ──────────────────  ────────────────────────────────────────
Production  main                https://example.com
Staging     staging             https://staging-123.myshopify.dev
Preview
```

So it looks like it's dangling off the side. Maybe if I added something instead of empty strings?

```
Name        Branch              URL
──────────  ──────────────────  ────────────────────────────────────────
Production  main                https://example.com
Staging     staging             https://staging-123.myshopify.dev
Preview     -                   -
```

Actually that kind of looks good!

Reply:
Ah I see. But what would be the "branch" that we need to specify in this case in `--branch`? I guess we just omit it?

Reply:
Exactly, `Preview` is the default. Right now environment names aren't unique so if you had multiple `Staging` branches there'd be no way to differentiate between them. Branches *are* unique on environments which is why we're using that field and making Preview default.